src/Common/src/System/Security/Cryptography/Asn1V2.cs

```diff
@@ -0,0 +1,5071 @@
+        private static bool TryCopyConstructedOctetStringValue(
+            ReadOnlyMemory<byte> source,
+            Span<byte> dest,
+            bool write,
+            AsnEncodingRules ruleSet,
+            bool isIndefinite,
+            out int bytesRead,
+            out int bytesWritten)
+        {
+            int lastSegmentSize = MaxCERSegmentSize;
+
+            Span<byte> tmpDest = dest;
+
+            int contentLength = CopyConstructedOctetString(
+                source,
+                ref tmpDest,
+                false,
+                ruleSet,
+                isIndefinite,
+                ref lastSegmentSize,
+                out int encodedLength);
+
+            // T-REC-X.690-201508 sec 9.2
+            if (ruleSet == AsnEncodingRules.CER && contentLength <= MaxCERSegmentSize)
+            {
+                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+
+            if (!write)
+            {
+                bytesRead = encodedLength;
+                bytesWritten = contentLength;
+                return true;
+            }
+
+            if (dest.Length < contentLength)
+            {
+                bytesRead = 0;
+                bytesWritten = 0;
+                return false;
+            }
+
+            tmpDest = dest;
+            lastSegmentSize = MaxCERSegmentSize;
+
+            bytesWritten = CopyConstructedOctetString(
+                source,
+                ref tmpDest,
+                true,
+                ruleSet,
+                isIndefinite,
+                ref lastSegmentSize,
+                out bytesRead);
+
+            Debug.Assert(bytesWritten == contentLength);
+            Debug.Assert(bytesRead == encodedLength);
+            return true;
+        }
+
+        public bool TryCopyOctetStringBytes(
+            Span<byte> destination,
+            out int bytesWritten)
+        {
+            return TryCopyOctetStringBytes(
+                Asn1Tag.PrimitiveOctetString,
+                destination,
+                out bytesWritten);
+        }
+
+        public bool TryCopyOctetStringBytes(
+            Asn1Tag expectedTag,
+            Span<byte> destination,
+            out int bytesWritten)
+        {
+            if (TryGetOctetStringBytes(
+                expectedTag,
+                out ReadOnlyMemory<byte> contents,
+                out int headerLength))
+            {
+                if (contents.Length > destination.Length)
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+
+                contents.Span.CopyTo(destination);
+                bytesWritten = contents.Length;
+                _data = _data.Slice(headerLength + contents.Length);
+                return true;
+            }
+
+            (Asn1Tag tag, int? length) = ReadTagAndLength(out headerLength);
+
+            bool copied = TryCopyConstructedOctetStringValue(
+                Slice(_data, heade
```

Comment:
Does the note in 8.8.2 ensure that the length of length is 1 octet? The note says "NOTE -- The length octet is zero" which implies a single octet to me.

Reply:
I agree that the note implies that.  But it uses a different wording than EOC does ("The end-of-contents octets shall consist of two zero octets." vs "The contents octets shall not contain any octets. (NOTE - The length octet is zero.)").

Checking with a few different implementations it seems that they're mostly okay with the multi-byte form.

`30 80 01 01 00 05 84 00 00 00 00 00 00` is a BER encoding of `SEQUENCE { (BOOLEAN) TRUE, NULL }`:

If you go to http://asn1-playground.oss.com/ and apply a schema of

```asn.1
Test DEFINITIONS IMPLICIT TAGS ::= 
BEGIN
  Data ::= SEQUENCE {
     a BOOLEAN,
     b NULL
  }
END
```

(then hit compile, then paste the hex string in the top-right box, then hit decode) it comes back as

```
Data SEQUENCE: tag = [UNIVERSAL 16] constructed; length = indef
  a BOOLEAN: tag = [UNIVERSAL 1] primitive; length = 1
    FALSE
  b NULL: tag = [UNIVERSAL 5] primitive; length = 0
    NULL
EOC
Successfully decoded 13 bytes.
rec1value Data ::= 
{
  a FALSE,
  b NULL
}
```
OpenSSL:
```
$ echo 30 80 01 01 00 05 84 00 00 00 00 00 00 | xxd -r -p | openssl asn1parse -i -dump -inform der
    0:d=0  hl=2 l=inf  cons: SEQUENCE
    2:d=1  hl=2 l=   1 prim:  BOOLEAN           :0
    5:d=1  hl=6 l=   0 prim:  NULL
   11:d=1  hl=2 l=   0 prim:  EOC
```

Windows:
```
D:\>echo MIABAQAFhAAAAAAAAA== > foo.ber

D:\>certutil -asn foo.ber
0000: 30 80                                     ; SEQUENCE (b Bytes - Indefinite Length)
0002: |  01 01                                  ; BOOL (1 Bytes)
0004: |  |  00
0005: |  05 84 00 00 00 00                      ; NULL (0 Bytes)
000b: |  00 00                                  ; Indefinite Length (@0000)
CertUtil: -asn command completed successfully.
```

Trying it without the wrapper sequence (just using the data `05 84 00 00 00 00`) makes OpenSSL fail to parse, but the other two are happy.

```asn.1
Test DEFINITIONS IMPLICIT TAGS ::= 
BEGIN
  Data ::= CHOICE {
    a NULL,
    b INTEGER
  }
END
```

So supporting multi-octet encoded zero-length for NULL seems to be in good company.

(On a whim, I checked the 1994 version of X.690, and the note was there, too.  So... strange.)